src/v/kafka/server/tests/metadata_test.cc

```diff
@@ -0,0 +1,336 @@
+        kafka::sasl_authenticate_request client_first_req;
+        {
+            auto msg = client_first.message();
+            client_first_req.data.auth_bytes = bytes(msg.cbegin(), msg.cend());
+        }
+        auto client_first_resp = client.dispatch(client_first_req).get0();
+        BOOST_REQUIRE_EQUAL(
+          client_first_resp.data.error_code, kafka::error_code::none);
+        return security::server_first_message(
+          client_first_resp.data.auth_bytes);
+    }
+
+    security::server_final_message send_scram_client_final(
+      kafka::client::transport& client,
+      const security::client_final_message& client_final) {
+        kafka::sasl_authenticate_request client_last_req;
+        {
+            auto msg = client_final.message();
+            client_last_req.data.auth_bytes = bytes(msg.cbegin(), msg.cend());
+        }
+        auto client_last_resp = client.dispatch(client_last_req).get0();
+
+        BOOST_REQUIRE_EQUAL(
+          client_last_resp.data.error_code, kafka::error_code::none);
+        return security::server_final_message(
+          std::move(client_last_resp.data.auth_bytes));
+    }
+
+    void do_sasl_handshake(kafka::client::transport& client) {
+        kafka::sasl_handshake_request req;
+        req.data.mechanism = security::scram_sha256_authenticator::name;
+
+        auto resp = client.dispatch(req).get0();
+        BOOST_REQUIRE_EQUAL(resp.data.error_code, kafka::error_code::none);
+    }
+
+    void authn_kafka_client(
+      kafka::client::transport& client,
+      const ss::sstring& username,
+      const ss::sstring& password) {
+        do_sasl_handshake(client);
+        const auto nonce = random_generators::gen_alphanum_string(130);
+        const security::client_first_message client_first(username, nonce);
+        const auto server_first = send_scram_client_first(client, client_first);
+
+        BOOST_REQUIRE(
+          std::string_view(server_first.nonce()).starts_with(nonce));
+        BOOST_REQUIRE_GE(
+          server_first.iterations(), security::scram_sha256::min_iterations);
+        security::client_final_message client_final(
+          bytes("n,,"), server_first.nonce());
+        auto salted_password = security::scram_sha256::hi(
+          bytes(password.cbegin(), password.cend()),
+          server_first.salt(),
+          server_first.iterations());
+        client_final.set_proof(security::scram_sha256::client_proof(
+          salted_password, client_first, server_first, client_final));
+
+        auto server_final = send_scram_client_final(client, client_final);
+        BOOST_REQUIRE(!server_final.error());
+
+        auto server_key = security::scram_sha256::server_key(salted_password);
+        auto server_sig = security::scram_sha256::server_signature(
+          server_key, client_first, server_first, client_final);
+
+        BOOST_REQUIRE_EQUAL(server_final.signature(), server_sig);
+    }
+};
```

Comment:
There's a lot duplication with `kafka/client/sasl_client`, was it too hard to extract?